### DISCUSSION TITLE
Extending TabItem doesn't work?

### DISCUSSION THREAD

**Discussion by irazasyed** [facebook/docusaurus#8864] [Q&A] [answered]
I'm trying to wrap TabItem into my own component with some default values but it doesn't seem to be working with the Tabs component.

Here's a CodeTabs component I'm working on:

```js
import React from 'react';
import Tabs from '@theme-original/Tabs';
import TabItem from '@theme/TabItem';
import clsx from 'clsx';
import LaravelLogo from '/img/laravel.svg';
import PHPLogo from '/img/php.svg';
import styles from './styles.module.css';

const languages = {
  php: { label: 'Standalone', logo: <PHPLogo className={clsx(styles.codeTabLogo, styles.logoPhp)} /> },
  laravel: { label: 'Laravel', logo: <LaravelLogo className={clsx(styles.codeTabLogo, styles.logoLaravel)} /> },
};

function Logo({ name }) {
  return (
    <div className={styles.codeTabLabel}>
      {languages[name].logo}
      <span>{languages[name].label}</span>
    </div>
  );
}

export function CodeTabs(props) {
  const values = Object.keys(languages).map(name => ({
    label: <Logo name={name} />,
    value: name,
    attributes: { className: styles[name] },
  }));

  return (
    <Tabs groupId='language' defaultValue='php' queryString key='language' values={values} {...props} />
  );
}

export function TabPHP({ children }) {
  return (
    <TabItem value='php'>

      {children}

    </TabItem>
  );
}

export function TabLaravel({ children }) {
  return (
    <TabItem value='laravel'>

      {children}

    </TabItem>
  );
}


```

The idea is to make it easier to use these tabs across all the docs. Example:

```js
import { CodeTabs, TabPHP, TabLaravel } from '@site/src/components/CodeTabs';

<CodeTabs>
<TabPHP>
SOME PHP ONLY CONTENT.
</TabPHP>
<TabLaravel>
LARAVEL SPECIFIC CONTENT.
</TabLaravel>
</CodeTabs>
```

Now it does show the tabs and changes the state BUT it does not hide the content nor switch between the content. However, If I use TabItem with the value set within the custom CodeTabs component, then it works fine.

What could be the issue with this? It doesn't let you extend or wrap?

**Comment by slorber** [comment, marked as answer]
This can't work because the Tabs component needs to see the props of his direct children's. 

Tabs need to "extract the config" without rendering the children:
- If you write `<TabPHP>` it can't see the value without running the function
- If you write `<TabItem value='php'>` it can extract the value: `children[0].props.value`

You can try building another API with these constraints in mind. 

If the number of languages is always 2, you can try:

```tsx
<CodeTabs>
  <>
    SOME PHP ONLY CONTENT.
  </>
  <>
    LARAVEL SPECIFIC CONTENT.
  </>
</CodeTabs>
```

You can also try:

```tsx
<CodeTabs 
  php={
    <>
      SOME PHP ONLY CONTENT.
    </>
  }
  laravel={
    <>
      SOME LARAVEL ONLY CONTENT.
    </>
  }
/>

```


TabItem can be used inside CodeTabs directly, the as long as they are direct child of Tabs that works.

You can also pass values directly as props to the parent Tabs component:

<img width="389" alt="CleanShot 2023-04-06 at 19 14 25@2x" src="https://user-images.githubusercontent.com/749374/230449343-40b7f178-9d1d-4ff2-a059-36538436f369.png">

Creating a repro that I can edit would be simpler for me to help you. Show me what's your goal with a mockup

**Reply by irazasyed** [reply]
Thanks for the detailed response. Very helpful. I'll try your ideas.

Here's the current component code for reference (I had to revert to using TabItem): [CodeTabs](https://github.com/telegram-bot-sdk/website/blob/main/src/components/CodeTabs/index.js).

It's going to be always 2 languages only. Here's the site (Linking to specific page for reference to the tabs usage): https://telegram-bot-sdk.com/docs/getting-started/configuration

**Reply by irazasyed** [reply]
I ended up refactoring to this:

```js
const tabItems = Children.map(props.children, (child, index) => {
    const language = Object.keys(languages)[index];

	return (
      <TabItem value={language}>
        {child}
      </TabItem>
    );
});

return (
    <Tabs
      groupId='language'
      defaultValue='php'
      queryString
      key='language'
      values={values}
      {...props}
    >
      {tabItems}
    </Tabs>
);
```

Usage:

```
<CodeTabs>
  <>
    SOME PHP ONLY CONTENT.
  </>
  <>
    LARAVEL SPECIFIC CONTENT.
  </>
</CodeTabs>
```
